### ISSUE TITLE
Shaping and instantiating variable fonts

### ISSUE THREAD

**Issue by simoncozens** [harfbuzz/harfbuzz#2112]
I seem to remember Behdad saying that Harfbuzz will not just be able to shape but also instantiate static instances of variable fonts. This will be helpful for me because, as a layout engine, SILE needs to pass fonts to a PDF generator, and the PDF library doesn't support variable fonts.

I just found the shaping example in the documentation (oops), but is the instantiator done yet? Is it on the radar?

**Comment by ebraminio** [comment]
This will be instatiation feature of subsetter as also referred on #1558 as a not in scope there, perhaps @blueshade7 and @garretrieger can comment on the status and planning for instatiation.

**Comment by garretrieger** [comment]
It's something that we want to add, but not being actively worked on at the moment.

**Comment by simoncozens** [comment]
Thanks, good to know. I’m happy to try working on this if someone can spec out an interface to it.

**Comment by ebraminio** [comment]
Perhaps Garret should comment here, hb_subset has a simple API https://github.com/harfbuzz/harfbuzz/issues/1381#issuecomment-439063823 so I guess var coords will become a new field on subset_input, if, not another API like hb_subset_var_instantiate is considered that takes a hb_font_t instead a face and returns instanced hb_face_t, which wouldn't need to use subset input even. It should be started by removing var tables and applying mvar to the needed places, and the real challenge will be in glyf/CFF/CFF2 and GSUB/GSUB/GPOS rebuild. I guess we can progress with a bottom up approach with adding related methods to each of those tables.

**Comment by khaledhosny** [comment]
I think one advantage of doing instanciation and sunsetting at the same time would be when instantiating large fonts to be embedded in PDF files, to avoid wasting time on glyphs that would be immediately sunsetted away.

**Comment by behdad** [comment]
We should add variation-space specification to subset input and cut the font to adhere to that.  That defaults to preserving all variation space, but should allow for narrowing down the space all the way to instantiating a single instance and leaving no variation behind.  So, @anthrotype's work on partial-instancing in fonttools should be considered at the same time, for API and implementation design purposes, even if not immediately implemented.

**Comment by simoncozens** [comment]
OK, I've been looking at this again; here are some things I think we will need to do.

- [ ] Introduce a `hb_variation_range_t` with lower and upper bounds.
- [ ] Add subsetting/serializing to tables we currently pass through: MVAR, avar, STAT, fvar.
- [ ] Properly deserialise into objects things we currently just treat as blobs and extract data when required: `TupleVariationHeader`, possibly others.
- [ ] Properly serialise a lot of things that we currently just copy: `Glyph`/`contour_point_vector_t`/`contour_point__t`.
- [ ] Add `hb_variation_range_t *variation_ranges; unsigned int variation_ranges_length;` to `hb_subset_plan_t`
- [ ] Subset/instance by applying all the variation ranges in gvar/glyf/MVAR/HVAR/VVAR/avar/STAT/fvar. (I don't care about CFF.)
- [ ] Terrible and scary things to GSUB/GPOS/GDEF. :-/

**Comment by twardoch** [comment]
FWIW, while I admire the goal of reducing varfonts to other varfonts (limiting the variation space), I think a sensible v1.0 goal would be subsetting to statics ie. producing a font that represents only a point in the variation space, not a region.

**Comment by simoncozens** [comment]
Meh. It's not actually that much more code. I'd *still* have to do all the things I mentioned.

**Comment by KrasnayaPloshchad** [comment]
Does HarfBuzz having some APIs that could be helpful  for this? If so, anyone could make use of them to subset a single instance, even if HarfBuzz doesn’t have such functionality.

**Comment by behdad** [comment]
This is starting to happen. First PR is https://github.com/harfbuzz/harfbuzz/pull/3708#pullrequestreview-1030514802
Closing as no point keeping open.

**Comment by papandreou** [comment]
Hey, we've recently implemented tracing of used variation axis ranges in [subfont](https://github.com/Munter/subfont), mainly in anticipation of this feature. We would like to do automatic instantiation/axis trimming to make it easier to author performant web pages with variable fonts. Just a :raised_back_of_hand: that we'd be excited to beta test whenever you have something ready! :hugs:

We currently use the subsetting features via `harfbuzzjs`.

**Comment by garretrieger** [comment]
Full instancing (that is reducing a variable font to a fully static instance) should be ready for use in the very near future. Mainly we're waiting on https://github.com/harfbuzz/harfbuzz/pull/3816 to be merged. Partial instancing (pinning/modifying some but not all axes) will be worked on next.

**Comment by khaledhosny** [comment]
Is instantiating CFF2 table supported or planned?

**Comment by garretrieger** [comment]
CFF2 is not currently supported. Currently the focus is on implementing for glyf. We do plan to also do instancing for CFF2 as well once the glyf implementation is done.

**Comment by papandreou** [comment]
We've just released [subfont 7.0.0](https://github.com/Munter/subfont/blob/master/CHANGELOG.md#v700-2023-04-08), which uses `hb_subset_input_pin_axis_location` to automatically perform a full instancing of variable fonts when possible. Looking forward to partial instancing :rocket:
